File: java-stream-mvn/src/main/java/OffsetTrackingReceive.java
import com.rabbitmq.stream.ByteCapacity;
import com.rabbitmq.stream.Environment;
import com.rabbitmq.stream.OffsetSpecification;

import java.time.Duration;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

import static java.nio.charset.StandardCharsets.UTF_8;

public class OffsetTrackingReceive {

  public static void main(String[] args) throws InterruptedException {
    try (Environment environment =
        Environment.builder().requestedHeartbeat(Duration.ofSeconds(5)).build()) {
      String stream = "stream-offset-tracking-java";
      environment.streamCreator().stream(stream).maxLengthBytes(ByteCapacity.GB(1)).create();

      OffsetSpecification offsetSpecification = OffsetSpecification.first();
      AtomicLong firstOffset = new AtomicLong(-1);
      AtomicLong lastOffset = new AtomicLong(0);
      AtomicLong messageCount = new AtomicLong(0);
      CountDownLatch consumedLatch = new CountDownLatch(1);
      environment.consumerBuilder().stream(stream)
          .offset(offsetSpecification)
          .name("offset-tracking-tutorial")
          .manualTrackingStrategy().builder()
          .messageHandler(
              (ctx, msg) -> {
                if (firstOffset.compareAndSet(-1, ctx.offset())) {
                  System.out.println("First message received.");
                }
                if (messageCount.incrementAndGet() % 10 == 0) {
                  ctx.storeOffset();
                }
                String body = new String(msg.getBodyAsBinary(), UTF_8);
                if ("marker".equals(body)) {
                  lastOffset.set(ctx.offset());
                  ctx.storeOffset();
                  ctx.consumer().close();
                  consumedLatch.countDown();
                }
              })
          .build();
      System.out.println("Started consuming...");

      consumedLatch.await(60, TimeUnit.MINUTES);

      System.out.printf("Done consuming, first offset %d, last offset %d.%n",
          firstOffset.get(), lastOffset.get());
    }
  }

}

File: dotnet-stream/OffsetTrackingReceive/OffsetTrackingReceive.cs
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using RabbitMQ.Stream.Client;
using RabbitMQ.Stream.Client.Reliable;

var streamSystem = await StreamSystem.Create(new StreamSystemConfig());

var stream = "stream-offset-tracking-dotnet";
await streamSystem.CreateStream(new StreamSpec(stream));

var consumerName = "offset-tracking-tutorial"; 
IOffsetType offsetSpecification;
try {
    ulong storedOffset = await streamSystem.QueryOffset(consumerName, stream).ConfigureAwait(false);
    offsetSpecification = new OffsetTypeOffset(storedOffset + 1);
} catch (OffsetNotFoundException) {
    offsetSpecification = new OffsetTypeFirst();
}
ulong initialValue = UInt64.MaxValue;
ulong firstOffset = initialValue;
int messageCount = 0;
ulong lastOffset = initialValue;
var consumedCde = new CountdownEvent(1);
var consumer = await Consumer.Create(new ConsumerConfig(streamSystem, stream)
{
    OffsetSpec = offsetSpecification,
    Reference =  consumerName,
    MessageHandler = async (_, consumer, context, message) => {
        if (Interlocked.CompareExchange(ref firstOffset, context.Offset, initialValue) == initialValue) {
            Console.WriteLine("First message received.");
        }
        if (Interlocked.Increment(ref messageCount) % 10 == 0) {
            await consumer.StoreOffset(context.Offset).ConfigureAwait(false);
        }
        if ("marker".Equals(Encoding.UTF8.GetString(message.Data.Contents))) {
            Interlocked.Exchange(ref lastOffset, context.Offset);
            await consumer.StoreOffset(context.Offset).ConfigureAwait(false);
            await consumer.Close();
            consumedCde.Signal();
        }
        await Task.CompletedTask;
    }
});
Console.WriteLine("Started consuming...");

consumedCde.Wait();
Console.WriteLine("Done consuming, first offset {0}, last offset {1}", firstOffset, lastOffset);
await streamSystem.Close();
